src/paths.ts

```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
     const serverIsWindows = /^\/[a-zA-Z]:\//.test(serverPath);
+
     if (serverIsWindows) {
         serverPath = serverPath.substr(1);
     }
+
+    if ( typeof pathMapping !== 'undefined' ) {
+        let mappedServerSource: string;
+
+        for (mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+
+            // normalize slashes for windows-to-unix
+            // I thought normalize did this automagically..
+            if ( process.platform !== 'win32' ) {
+                mappedServerSource = mappedServerSource.replace(/\\/g, '/');
+            }
+
+            let serverRelative: string = path.relative(mappedServerSource, serverPath);
+
+            if ( serverRelative && serverRelative.length > 0 ) {
+                let relative: number = serverRelative.indexOf('..');
+
+                // if path does not start with ..
+                if ( relative !== 0 ) {
+                    serverSourceRoot = mappedServerSource;
+                    localSourceRoot = mappedLocalSource;
+
+                    break;
+                }
+            }
+        }
+    }
+
     let localPath: string;
-    if (serverSourceRoot && localSourceRoot) {
+
+    if (typeof serverSourceRoot !== 'undefined' && serverSourceRoot && typeof localSourceRoot !== 'undefined' && localSourceRoot) {
         // get the part of the path that is relative to the source root
         const pathRelativeToSourceRoot = (serverIsWindows ? path.win32 : path.posix).relative(serverSourceRoot, serverPath);
+
         // resolve from the local source root
         localPath = path.resolve(localSourceRoot, pathRelativeToSourceRoot);
+
     } else {
         localPath = path.normalize(serverPath);
     }
+
     return localPath;
 }
 
 /** converts a local path from VS Code to a server-side XDebug file URI with respect to source root settings */
-export function convertClientPathToDebugger(localPath: string, localSourceRoot?: string, serverSourceRoot?: string): string {
+export function convertClientPathToDebugger(localPath: string, localSourceRoot?: string, serverSourceRoot?: string, pathMapping?: { [index: string]: string; }): string {
+
     let localFileUri = fileUrl(localPath, {resolve: false});
     let serverFileUri: string;
-    if (serverSourceRoot && localSourceRoot) {
+
+    if ( typeof pathMapping !== 'undefined' ) {
+        let mappedServerSource: string;
+
+        for (mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+
+            let localRelative: string = path.relative(mappedLocalSource, localPath);
+
+            if ( localRelative && localRelative.length > 0 ) {
+                let relative: number = localRelative.indexOf('..');
+
+                // if does not start with ..
+                if ( relative !== 0 ) {
```

Comment:
this is not used anywhere else, so no reason to save it in a variable. You can also use `.includes()` if I'm not mistaken.